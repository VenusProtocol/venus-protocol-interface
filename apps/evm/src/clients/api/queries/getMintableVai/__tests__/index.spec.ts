import type { Address, PublicClient } from 'viem';

import fakeAccountAddress from '__mocks__/models/address';

import { getMintableVai } from '..';

const fakeVaiAddress: Address = '0xfakeVaiAddress';
const fakeVaiControllerContractAddress: Address = '0xfakeVaiControllerContractAddress';

describe('getMintableVai', () => {
  test('returns the mintable VAI amount on success', async () => {
    const fakeMintCapMantissa = 1000000000000000n;
    const fakeTotalSupplyMantissa = 100000000000000n;
    const fakeMintableVaiMantissa = 100000000000000n;

    const multicallMock = vi.fn(() => [
      {
        result: fakeMintCapMantissa,
        status: 'success',
      },
      {
        result: fakeTotalSupplyMantissa,
        status: 'success',
      },
      {
        result: [undefined, fakeMintableVaiMantissa],
        status: 'success',
      },
    ]);

    const fakePublicClient = {
      multicall: multicallMock,
    } as unknown as PublicClient;

    const response = await getMintableVai({
      publicClient: fakePublicClient,
      vaiAddress: fakeVaiAddress,
      vaiControllerContractAddress: fakeVaiControllerContractAddress,
      accountAddress: fakeAccountAddress,
    });

    expect(multicallMock).toHaveBeenCalledTimes(1);
    expect(multicallMock).toHaveBeenCalledWith({
      contracts: [
        {
          abi: expect.any(Object),
          address: fakeVaiControllerContractAddress,
          functionName: 'mintCap',
        },
        {
          abi: expect.any(Object),
          address: fakeVaiAddress,
          functionName: 'totalSupply',
        },
        {
          abi: expect.any(Object),
          address: fakeVaiControllerContractAddress,
          functionName: 'getMintableVAI',
          args: [fakeAccountAddress],
        },
      ],
    });

    expect(response).toMatchSnapshot();
  });
});
