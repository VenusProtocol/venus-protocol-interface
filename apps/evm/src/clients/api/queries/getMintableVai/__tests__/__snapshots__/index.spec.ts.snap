// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getMintableVai > returns the mintable VAI amount on success 1`] = `
{
  "accountMintableVaiMantissa": "100000000000000",
  "vaiLiquidityMantissa": "900000000000000",
}
`;
